keps/sig-auth/3299-kms-v2-improvements/README.md

```diff
@@ -337,57 +303,96 @@ sequenceDiagram
+[x] I/we understand the owners of the involved components may require updates to existing tests to make this code solid enough prior to committing the changes necessary to implement this enhancement.
 
 ##### Prerequisite testing updates
 
-This section is incomplete and will be updated before the beta milestone.
-
 ##### Unit tests
 
-This section is incomplete and will be updated before the beta milestone.
+- Validate key hierarchy behavior in reference implementation
+  - Local KEK is generated for encryption
+  - Local KEK is rotated after a period of time/number of operations
+  - Remote KEK is only called when
+    - Cache is empty, so remote KEK is needed to decrypt local KEK
+    - Local KEK is generated or rotated
+  - When key hierarchy is not used, remote KEK is called for every encryption/decryption
+  - Ensure the logs and metrics are generated as expected
+  - At least 75% code coverage
+- Staleness check based on keyID in the `StatusResponse`
+- Unit test for gRPC request/response validation
+  - Serialize/Deserialize `EncryptedObject`
+  - Validate keyID in `StatusResponse` and `EncryptResponse`
+  - Validate annotations in `EncryptResponse`
+  - Validate logs are generated with the correct `UID` and additional metadata
 
 ##### Integration tests
 
-This section is incomplete and will be updated before the beta milestone.
+- Integration tests to validate
+  - Encryption of custom resources and custom resource definitions
+  - No-op writes cause rewrite of stale data (data that has correct schema but was encrypted with keyID that is not the latest)
+  - Health checks
+    - single health check for v2 at `/kms-providers`
+    - individual health checks for v1 and v2 with `/kms-provider-0` and `/kms-provider-1`
+- Integration tests with base64 plugin to validate the encryption and decryption of data
+- Integration tests with reference implementation
+  - Validate the encryption and decryption of data
+  - Validate the functionality of reference implementation
+- Integration tests to check rotation is possible without restarting API server
+- Integration tests that exercise the feature enablement/disablement flow
 
 ##### e2e tests
 
-This section is incomplete and will be updated before the beta milestone.
+With this e2e test suite, we want to do the following:
+1. Run the e2e suite against a kind cluster without kms encryption enabled.
+2. Run the e2e suite against a kind cluster that has kms v2 encryption enabled (as defined below).
+3. Compare `request_duration_seconds`, `request_terminations_total`, `request_aborts_total` API server metrics between the two runs. The acceptable delta should be less than 20%.
+4. Observe metrics from the reference implementation to determine time taken at each step of the encryption/decryption process.
+5. Observe API server startup time with and without kms encryption enabled.
```

Comment:
Since we defined the acceptable delta for API server metrics (latency) to be 20% we can do the same here. WDYT @enj?

Reply:
@logicalhan Me and @enj had a conversation and this is the rough math we came up with:

Assuming 10ms for every KMS RPC request to be complete, for every 5000 resource encrypted, it will increase the startup time by roughly a minute (5000*10ms = 50s). Is this something you would like us to documented in the KEP?